### ISSUE TITLE
Concurrency Overhaul for RepoSense

### ISSUE THREAD

**Issue by yhtMinceraft1010X** [reposense/RepoSense#1669]
**Related to issue #1601 and PR #1625**

Although I have a PR for migrating to thread-safe date-time APIs, after some research, I believe that this may not fully solve the concurrency issue that was exposed in #1601. 

**Resources on other non-thread-safe APIs:**

* https://www.quora.com/What-are-some-commonly-used-Java-API-classes-that-are-not-thread-safe
* https://stackoverflow.com/questions/3362018/is-linkedlist-thread-safe-when-im-accessing-it-with-offer-and-poll-exclusively#:~:text=LinkedList%20is%20not%20thread%20safe,slightly%20different%20behavior%20than%20LinkedList.
* https://stackoverflow.com/questions/1360113/is-java-regex-thread-safe
* https://www.quora.com/Is-ArrayList-thread-safe
* https://stackoverflow.com/questions/6045648/which-java-collections-are-synchronizedthread-safe-which-are-not#:~:text=All%20collection%20classes%20(except%20Vector,%2Dsafe%3A%20Vector%20and%20Hashtable.

**In short, these are the classes that we may need to change to fully resolve any concurrency issues:**

* `SimpleDateFormat` and `Calendar` - PR #1625 should address this once merged
* `Matcher` - From the [java regex thread safety post](https://stackoverflow.com/questions/1360113/is-java-regex-thread-safe), it appears there is a way to use `Matcher` in a thread-safe manner
* `HashMap` - Consider changing to [`ConcurrentHashMap`](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/package-summary.html)
* `ArrayList` - Consider changing to [`CopyOnWriteArrayList`](https://docs.oracle.com/javase/8/docs/api/index.html?java/util/concurrent/package-summary.html)

Many of the above classes are used in code to analyse authorship and commit info, which uses multiple threads in `cloneAndAnalyzeRepos` method of `ReportGenerator`.

**The following classes are confirmed to be thread-safe:**

* [`Path`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html)
* [`Logger`](https://docs.oracle.com/javase/8/docs/api/java/util/logging/Logger.html)

**Recommendation:**

Before jumping into solving any of these, we should spend some time to look through the `src` code, find any other classes/implementations that are potentially not thread-safe, consider the best thread-safe alternative to use and then update this issue.

**Comment by chan-j-d** [comment]
From my experience, the concurrent collections are noticeably slower in performance than their non-thread safe counterparts. Also I'm not too sure if there are that many asynchronous tasks in RepoSense right now that can cause race conditions. The main asynchronous tasks that I am aware of are the cloning and analyzing done here:

https://github.com/reposense/RepoSense/blob/ed131f6294bf83fac436a6265bc4d53b022739f7/src/main/java/reposense/report/ReportGenerator.java#L227-L234

Each is done on a single repo/repo configuration so it seems like there would not be a lot of shared variables for concurrency issues. Within each task, everything is done synchronously so the non-thread safe constructs should be fine. 

Having said that, there is this bug which occasionally causes gradle build to fail: https://github.com/reposense/RepoSense/pull/1664#issuecomment-1038208378. that might possibly be due to race conditions as well.

**Comment by yhtMinceraft1010X** [comment]
I also encountered something similar to that error in one [workflow run](https://github.com/reposense/RepoSense/runs/5186440334?check_suite_focus=true), unlike the one in #1664, this one happens in testrepo-Delta instead of testrepo-Charlie.

![Screenshot (106)](https://user-images.githubusercontent.com/69678785/154019961-56c152f5-8010-472a-b306-b930d6b364f7.png)
Here's a picture of the differences between expected and actual. I noticed that some dates are off:
* The dates look weird (like year `0012`, `1414`)
* The dates look correct but don't seem to fit anywhere (e.g. the 2017 one).

Upon closer inspection for `lithiumlkid`, I noticed that for `2018-04-13`, the 3 commits are present for both expected and actual, but are ordered wrongly:
![Ordering](https://user-images.githubusercontent.com/69678785/154022306-684e5221-0205-4eb8-9b59-e1f9b72a3544.png)

Come to think of it, this may all be due to `SimpleDateFormat`. For the 3 commits on `2018-04-13`, it's probably because the exact time was parsed incorrectly (but time is not shown in json) and that affected the sorting in `CommitInfoAnalyzer.analyzeCommits()`. 

Reference commit list: https://github.com/reposense/testrepo-Delta/commits?author=ahmadsyafiqruhazat (April 13 2018)
